src/dune_rules/melange/melange_rules.ml

```diff
@@ -16,36 +16,37 @@ let ocaml_flags sctx ~dir melange =
     let ocaml_version = (Super_context.context sctx).version in
     Super_context.with_vendored_flags ~ocaml_version flags
 
-let lib_output_dir ~sctx ~target_dir lib =
+let output_of_lib ~target_dir lib =
   let info = Lib.info lib in
-  let lib_dir =
-    match Lib_info.status info with
-    | Private _ ->
-      let lib = Lib.Local.of_lib_exn lib in
-      let info = Lib.Local.info lib in
-      Lib_info.src_dir info
-    | Public _ ->
-      let package_name = Option.value_exn (Lib_info.package info) in
-      let bctx = (Super_context.context sctx).build_dir in
-      let info = Lib.info lib in
-      let src_dir = Lib_info.src_dir info in
-      Path.Build.L.relative bctx
-        [ "node_modules"
-        ; Package.Name.to_string package_name
-        ; Path.Source.to_string (Path.drop_build_context_exn src_dir)
-        ]
-    | Installed | Installed_private ->
-      let package_name = Option.value_exn (Lib_info.package info) in
-      let bctx = (Super_context.context sctx).build_dir in
-      Path.Build.L.relative bctx
-        [ "node_modules"; Package.Name.to_string package_name ]
-  in
-  Path.Build.append_source target_dir
-    (Path.Build.drop_build_context_exn lib_dir)
+  match Lib_info.status info with
+  | Private _ -> `Private_library_or_emit target_dir
+  | Installed | Installed_private ->
+    let package_name = Option.value_exn (Lib_info.package info) in
+    `Public_library
+      (Path.Build.L.relative target_dir
+         [ "node_modules"; Package.Name.to_string package_name ])
+  | Public _ ->
+    let package_name = Option.value_exn (Lib_info.package info) in
+    let src_dir = Lib_info.src_dir info in
+    `Public_library
+      (Path.Build.L.relative target_dir
+         [ "node_modules"
+         ; Package.Name.to_string package_name
+         ; Path.Source.to_string (Path.drop_build_context_exn src_dir)
+         ])
 
-let make_js_name ~js_ext ~dst_dir m =
+let make_js_name ~js_ext ~output m =
   let name = Melange.js_basename m ^ js_ext in
-  Path.Build.relative dst_dir name
+  match output with
+  | `Public_library output_dir -> Path.Build.relative output_dir name
+  | `Private_library_or_emit target_dir ->
+    let dst_dir =
+      Path.Build.append_source target_dir
+        (Module.file m ~ml_kind:Impl
+        |> Option.value_exn |> Path.as_in_build_dir_exn |> Path.Build.parent_exn
```

Comment:
As this is chaining 4 functions that can raise, I wonder if any of the cases could ever happen and should be handled (maybe `parent_exn`?). Or are they all absolutely safe?

Reply:
In this case they're all safe:

1. we checked the module has an implementation
2. the module in question must always be from a private library or an emit stanza
3. since it's a file, it always has a parent.
4. There's only 3 functions :)

Reply:
I was counting `drop_build_context_exn` in the line below :)

Reply:
Yes, that one also is safe. That function only fails when the path is `_build/install`. But that's impossible for source paths of modules.